flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java

```diff
@@ -0,0 +1,408 @@
+	private final StatusWatermarkValve statusWatermarkValve2;
+
+	/**
+	 * Stream status for the two inputs. We need to keep track for determining when
+	 * to forward stream status changes downstream.
+	 */
+	private StreamStatus firstStatus;
+	private StreamStatus secondStatus;
+
+	private int availableInputsMask;
+
+	private int lastReadInputIndex;
+
+	private InputSelection inputSelection;
+
+	private Counter numRecordsIn;
+
+	private boolean initialized;
+
+	public StreamTwoInputSelectableProcessor(
+		Collection<InputGate> inputGates1,
+		Collection<InputGate> inputGates2,
+		TypeSerializer<IN1> inputSerializer1,
+		TypeSerializer<IN2> inputSerializer2,
+		Object lock,
+		IOManager ioManager,
+		StreamStatusMaintainer streamStatusMaintainer,
+		TwoInputStreamOperator<IN1, IN2, ?> streamOperator,
+		WatermarkGauge input1WatermarkGauge,
+		WatermarkGauge input2WatermarkGauge) {
+
+		checkState(streamOperator instanceof InputSelectable);
+
+		this.streamOperator = checkNotNull(streamOperator);
+		this.inputSelector = (InputSelectable) streamOperator;
+
+		this.lock = checkNotNull(lock);
+
+		InputGate unionedInputGate1 = InputGateUtil.createInputGate(inputGates1.toArray(new InputGate[0]));
+		InputGate unionedInputGate2 = InputGateUtil.createInputGate(inputGates2.toArray(new InputGate[0]));
+
+		// create a Input instance for each input
+		this.input1 = new StreamTaskNetworkInput(new BarrierDiscarder(unionedInputGate1), inputSerializer1, ioManager, 0);
+		this.input2 = new StreamTaskNetworkInput(new BarrierDiscarder(unionedInputGate2), inputSerializer2, ioManager, 1);
+
+		this.statusWatermarkValve1 = new StatusWatermarkValve(
+			unionedInputGate1.getNumberOfInputChannels(),
+			new ForwardingValveOutputHandler(streamOperator, lock, streamStatusMaintainer, input1WatermarkGauge, 0));
+		this.statusWatermarkValve2 = new StatusWatermarkValve(
+			unionedInputGate2.getNumberOfInputChannels(),
+			new ForwardingValveOutputHandler(streamOperator, lock, streamStatusMaintainer, input2WatermarkGauge, 1));
+
+		this.firstStatus = StreamStatus.ACTIVE;
+		this.secondStatus = StreamStatus.ACTIVE;
+
+		this.availableInputsMask = (int) new InputSelection.Builder().select(1).select(2).build().getInputMask();
+
+		this.lastReadInputIndex = 1; // always try to read from the first input
+
+		this.initialized = false;
+
+	}
+
+	public boolean processInput() throws Exception {
+		if (!initialized) {
+			initialize();
+		}
+
+		int readingInputIndex = inputSelection.fairSelectNextIndexOutOf2(availableInputsMask, lastReadInputIndex);
+		if (readingInputIndex == -1) {
+			return waitForAvailableInput(inputSelection);
+		}
+		lastReadInputIndex = readingInputIndex;
+
+		// if the input selection is ALL and availableInputsMask is not ALL,
+		// always try to check and set the availability of another input
+		if (availableInputsMask < 3 && inputSelection.isALLMaskOf2()) {
+			checkAndSetAvailable(1 - readingInputIndex);
```

Comment:
Also I think we should also add another benchmark. With `InputSelection.ANY`, but where one side of the input is  not available. However I'm not sure how we could easily implement such benchmark? As long as we are assuming that this is a micro benchmark, we might be able to implement it with two fake inputs, that communicate through some static variable/latch.

1st input has only one record to emit, but it sleeps/waits for 2nd input to complete first
2nd input has N records to emit. After emitting all of them, it releases the latch of the first input.

?

Reply:
Your idea is good and easy to implement. I've created the PR#22 (https://github.com/dataArtisans/flink-benchmarks/pull/22) for benchmarks.